test/hotspot/gtest/nmt/test_homogenousObjectArray.cpp

```diff
@@ -0,0 +1,152 @@
+struct LL {
+  struct Node;
+  using NodeAllocator = HomogenousObjectArray<Node, mtTest>;
+  using NodePtr = typename NodeAllocator::I;
+  NodeAllocator alloc;
+  struct Node {
+    E e;
+    NodePtr next;
+  };
+
+  NodePtr start;
+  LL()
+  : start{NodeAllocator::nil} {
+  }
+
+  void push(E e) {
+    NodePtr new_element = alloc.allocate(e, NodeAllocator::nil);
+    NodePtr& current = start;
+    if (current == NodeAllocator::nil) {
+      current = new_element;
+      return;
+    }
+    alloc.at(new_element).next = current;
+    current = new_element;
+  };
+
+  E pop() {
+    assert(start != NodeAllocator::nil, "must be");
+    Node& n = alloc.at(start);
+    E e = n.e;
+    NodePtr next_start = n.next;
+    alloc.deallocate(start);
+    start = next_start;
+    return e;
+  }
+};
+// A linked list which is capable of having multiple different allocators. This is done through higher-kinded types.
+// That's a very fancy word that means that a templated type like Foo<E> can be passed around like only Foo at first
+// and then be 'applied' to some E. Think of it like passing around a lambda or function pointer, but on a template level,
+// where Foo is a function that can be called on some type with the return type being Foo<E>.
+template<typename E, template<typename, MEMFLAGS> class Allocator>
+struct LL2 {
+  struct Node;
+  using NodeAllocator = Allocator<Node, mtTest>;
+  using NodePtr = typename NodeAllocator::I;
+  NodeAllocator alloc;
+  struct Node {
+    E e;
+    NodePtr next;
+  };
+
+  NodePtr start;
+  LL2()
+    : start(NodeAllocator::nil) {
+  }
+
+  void push(E e) {
+    NodePtr new_element = alloc.allocate(e, NodeAllocator::nil);
+    NodePtr& current = start;
+    if (current == NodeAllocator::nil) {
+      current = new_element;
+      return;
+    }
+    alloc.at(new_element).next = current;
+    current = new_element;
+  };
+
+  E pop() {
+    assert(start != NodeAllocator::nil, "must be");
+    Node& n = alloc.at(start);
+    E e = n.e;
+    NodePtr next_start = n.next;
+    alloc.deallocate(start);
+    start = next_start;
+    return e;
+  }
+};
+
+template<typename List>
+void test_with_list(List& list) {
+  list.push(1);
+  list.push(2);
+  EXPECT_EQ(2, list.pop());
+  EXPECT_EQ(1, list.pop());
+}
+
+TEST_VM_F(HomogenousObjectArrayTest, TestLinkedLists) {
+  {
+    LL<int> list;
+    test_with_list(list);
+  }
+  {
+    LL2<int, HomogenousObjectArray> list;
+    test_with_list(list);
+  }
+}
+
+TEST_VM_F(HomogenousObjectArrayTest, FreeingShouldReuseMemory) {
+  A alloc;
+  A::I i = alloc.allocate(1);
+  int* x = &alloc.at(i);
+  alloc.deallocate(i);
+  i = alloc.allocate(1);
+  int* y = &alloc.at(i);
+  EXPECT_EQ(x, y);
+}
+
+TEST_VM_F(HomogenousObjectArrayTest, FreeingInTheMiddleWorks) {
+  A alloc;
+  A::I i0 = alloc.allocate(0);
+  A::I i1 = alloc.allocate(0);
+  A::I i2 = alloc.allocate(0);
+  int* p1 = &alloc.at(i1);
+  alloc.deallocate(i1);
+  A::I i3 = alloc.allocate(0);
+  EXPECT_EQ(p1, &alloc.at(i3));
+}
```

Comment:
I think for what little the tests do, testing with different list types is overengineered. You could just scrap both LL and LL2, and just use a HOA directly.

Reply:
Hmm. I am not sold on the "example" benefit. The array is quite easy in itself.

You could scratch the lists, since their implementations don't do anything to really test the AWFL (new acronym, yay). For the saved LOCs could expand the tests to test with a collection of various types, e.g.
- u1, u2, u8
- unaligned structures that need alignment (e.g. struct (void*; int; ))
- trivial objects

Interesting to see would be that it works, that alignment is correct, that nothing breaks across resizes (indexes stay stable, etc.).

I wont insist of it, just thinking that the current test complexity is somewhat wasted.

Reply:
>really test the AWFL (new acronym, yay).

Can't wait for templatized indices so I can write

```c++
using AWFUL = ArrayWithFreeListAllocator<int, uint8_t>;
```

>Interesting to see would be that it works, that alignment is correct, that nothing breaks across resizes (indexes stay stable, etc.).

>I wont insist of it, just thinking that the current test complexity is somewhat wasted.

I think those tests make sense if we replace the `GrowableArray`as the backing memory area with something else, otherwise that's just testing the GA.

OK, how about we scrap the tests for when moving this to utilities and having the index being a part of the template? I've got that in the pipeline already, so it's not going to take too long I hope.

Reply:
Fine for me